### ISSUE TITLE
React Native server stories list infinitely loading

### ISSUE THREAD

**Issue by bouncycatt** [storybookjs/react-native#55]
Hello! I had a problem with loading Storybook web explorer, it doesn't load the left menu. But on a device Storybook works well. I met the problem both during Intro React Native Storybook Tutorial and in my bare workflow project. It just doesn't build a web version. 

**Steps**
1. In a terminal to a project folder:
`npx -p @storybook/cli sb init --type react_native`
`yarn add -D @storybook/addon-ondevice-actions`

2. In storybook/rn-addons.js
`import '@storybook/addon-ondevice-actions/register';`

3. In the terminal I tried this and reverse order:
`yarn storybook`
`yarn start`
`a` – both Android simulator and a real Android device, Storybook works well on both options but does not in a browser.

4. On `yarn storybook` the answer:
```
yarn run v1.22.4
$ start-storybook -p 7007
info => Loading presets
info => Loading custom manager config.
╭────────────────────────────────────────────────────╮
│                                                    │
│   Storybook 5.3.18 started                         │
│   4.39 s                                           │
│                                                    │
│    Local:            http://localhost:7007/        │
│    On your network:  http://192.168.0.104:7007/    │
│                                                    │
╰────────────────────────────────────────────────────╯
```

**From the tutorial it should be this**
<img width="356" alt="from tutorial" src="https://user-images.githubusercontent.com/63786160/82722658-62ecb780-9cd1-11ea-97ad-5e7a3d04ff3c.png">

**But I have this result on any projects and environments I tried**
<img width="843" alt="error" src="https://user-images.githubusercontent.com/63786160/82723362-12785880-9cd7-11ea-87e9-a2142f46cbe9.png">

**Environment Info:**
  System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
  Binaries:
    Node: 12.16.3 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
  Browsers:
    Chrome: 81.0.4044.138
    Firefox: 76.0
    Safari: 12.1.2
  npmPackages:
    @storybook/addon-actions: ^5.3.18 => 5.3.18 
    @storybook/addon-links: ^5.3.18 => 5.3.18 
    @storybook/addon-ondevice-actions: ^5.3.18 => 5.3.18 
    @storybook/addons: ^5.3.18 => 5.3.18 
    @storybook/react-native: ^5.3.18 => 5.3.18 
    @storybook/react-native-server: ^5.3.18 => 5.3.18 

Also, I tried with the older 5.2.1 Storybook version, an Android simulator, and an Android phone via USB, Ubuntu, manual server setup, but the same error. 

I also set up React Storybook Tutorial for a try to see if there is this error there. No problem, all works properly with React.

**Comment by JeffGuKang** [comment]
Did you try it? 
1. Connect Android device
2. Run storybook server, you need to check `$ (adb reverse tcp:7007 tcp:7007 || true) && start-storybook` working well. 
3. Run react native to your device
4. Reload your app in debug mode

**Comment by akhudairy** [comment]
Hi @bouncycatt .. How do you run the story book on the mobile or simulator? I have the same problem on web version and I did follow the tutorial steps but I feel something is missing.
For Android and IOS .. when I run "yarn ios" for example it runs the app on the ios but not the story book, so am not sure which commands runs the storybook on the device.

**Comment by abdullahizzuddiin** [comment]
Hi,
anyone has solved this issue?

I got same problem too.

**Comment by bouncycatt** [comment]
@JeffGuKang Thanks. I tried it but the problem is the same.

**Comment by bouncycatt** [comment]
@akhudairy Hi! If I'm not mistaken, a Storybook viewer appeared on a device when I tried previous versions of Storybook. Unfortunately, I don't remember. Today I retook all the tutorial steps with `abd` command but nothing happened.

**Comment by Esxiel** [comment]
@akhudairy You have to render the StorybookUIRoot. There's a part in https://storybook.js.org/docs/guides/guide-react-native/ where it says

> ... StorybookUI is a RN View component that can be embedded anywhere in your RN application, e.g. on a tab or within an admin screen.

This component is the "storybook on the mobile". Usually I'd have my entry point index.js's AppRegistry render out this component instead of the app component(using an if conditional check in regards to .env). Some people put it on a different screen in their navigation, etc.

**Comment by Esxiel** [comment]
@bouncycatt I'm using storybook on iOS so I can't comment much but my web storybook always looks like this when my app(on the simulator) does not touch/run the getStorybookUI function. This happened to me initially because I didn't know I had to import and render it somewhere in my app(see the comment I made above). 

I'm not 100% sure if this is the cause since I don't really know the inner workings on getStorybookUI/ how our app connects to the storybook react-native-server.

**Comment by akhudairy** [comment]
hi @Esxiel ... thanks i will give that ago by adding storybook to my other project using the manual steps.

But the idea is I was following the story book official React Native tutorial. and their app.js file looks very different compared to these manual steps ... here is the code they ask you to add

// App.js
async function loadResourcesAndDataAsync() {
  try {
    SplashScreen.preventAutoHide();

    // Load our initial navigation state
    setInitialNavigationState(await getInitialState());

    // Load fonts
    await Font.loadAsync({
      ...Ionicons.font,
      'space-mono': require('./assets/fonts/SpaceMono-Regular.ttf'),
      percolate: require('./assets/icon/percolate.ttf'),
      'NunitoSans-Bold': require('./assets/font/NunitoSans-Bold.ttf'),
      'NunitoSans-Italic': require('./assets/font/NunitoSans-Italic.ttf'),
      NunitoSans: require('./assets/font/NunitoSans-Regular.ttf'),
    });
  } catch (e) {
    // We might want to provide this error information to an error reporting service
    console.warn(e);
  } finally {
    setLoadingComplete(true);
    SplashScreen.hide();
  }
}

**Comment by Esxiel** [comment]
Um @akhudairy , I might be misunderstanding you, but you are referring to https://www.learnstorybook.com/intro-to-storybook/react-native/en/get-started/ right?

You are right in saying that the code in manual steps guide is very different with the App.js in the tutorial, however you'd benefit from understanding a few misconceptions you have...
1. The part of App.js you quote is about is solely about loading up fonts and hiding the splashscreen after it's done, it has no relation with implementing Storybook, as they said, it is to `To match the intended design, you'll need to download both the font and icon directories and place them inside the assets folder.`Also, remember that this tutorial of theirs rely on using that template project from expo, it's not really a tutorial meant to show you how to put storybook in your own project.
2. As I stated in my previous post, I had my entry point directly render out StorybookUI, however some people render the StorybookUI within a different screen inside a navigator, this is what the react native storybook tutorial does. The part from `In order to see Storybook in React Native we're going to update screens/LinksScreen.js to the following: ...` there on.

I apologize to the maintainers for veering off-topic.

**Comment by dannyhw** [comment]
Was anyone able to find a workaround? or the config that is needed to fix this.

**Comment by dannyhw** [comment]
You should run storybook server and then after run the react native app then if you refresh the webpage for react native server the stories should load.

**Comment by rmushtaq21** [comment]
Has anyone overcome this issue yet? I'm experiencing the same problem (forever loading left panel) as @bouncycatt when I try to run Storybook with **react-native-web**.

**Comment by allen-hsu** [comment]
I have the same issue, but, I resolve that, plz check your rn storybook port and storybook port are the same, in my case, 
default storybook server port is 6006, I change to 7007, or you can follow those document: 
https://github.com/storybookjs/react-native#storybook-for-react-native

getStorybookUI Options
port: Number (7007)
        -- port to use

**Comment by dannyhw** [comment]
I found there is an issue when using a real android device with expo. It can be resolved by specifying your port and host explicitly

When you start up the server you should see something with your IP here
On your network:  http://xxx.xxx.x.xx:7007/

Take this IP and set it explicitly as the host and then the storybookui should be able to reach the server

In code

```
const StorybookUIRoot = getStorybookUI({
  host: "192.111.1.11", // replace this ip address with your local ip address
  port: "7007"
});
```

In package.json `"storybook": "start-storybook -p 7007 -h 192.111.1.11"` (replace this ip address with your local ip address
  port: "7007")


This should resolve the connection issues

This might also help some people if the above isn't working

`adb reverse tcp:7007 tcp:7007`

**Comment by chadmott** [comment]
having the same issue... following the tutorial (twice) exactly as written, and the stories are perpetually loading for react native. 

React tutorial is working fine.

**Comment by dannyhw** [comment]
@chadmott are you using a simulator or a real device or web?

The tutorial is needing some updates so it might not be 100% accurate. Also the auto setup has a bug currently regarding dependency versions.

**Comment by dannyhw** [comment]
@chadmott check your package.json if there are any storybook packages with version 6.0 or above please downgrade them to 5.3.21 since we don't yet support storybook v6 with react native.

**Comment by chadmott** [comment]
thanks @dannyhw but that did not solve my issue. I'm just trying to learn Storybook (hence the tutorial) so I'm afraid I don't know why its broken. For now I'll just use the react tutorial

**Comment by dannyhw** [comment]
@chadmott the part where it says to use "yarn start" should be "yarn web" that should fix the issue.

edit:
Actually the server will continue to show the loading bars but actually that is fine because the stories will only get loaded once you do the later steps in the tutorial. This is because the storybookUI needs to be loaded for the server to receive the stories list.

**Comment by impactro** [comment]
ual, this is incredible, it worked now, thank you very much for your help.
I'm using Android emulator BlueStacks, all right now!
thank you very much for your help

![image](https://user-images.githubusercontent.com/1304804/92603750-103a0a00-f286-11ea-8ddc-05006f0cdad4.png)

**Comment by dannyhw** [comment]
@impactro great to see you were able to get it working :)

**Comment by Shhhhhreyas** [comment]
> Hello! I had a problem with loading Storybook web explorer, it doesn't load the left menu. But on a device Storybook works well. I met the problem both during Intro React Native Storybook Tutorial and in my bare workflow project. It just doesn't build a web version.
> 
> **Steps**
> 
> 1. In a terminal to a project folder:
>    `npx -p @storybook/cli sb init --type react_native`
>    `yarn add -D @storybook/addon-ondevice-actions`
> 2. In storybook/rn-addons.js
>    `import '@storybook/addon-ondevice-actions/register';`
> 3. In the terminal I tried this and reverse order:
>    `yarn storybook`
>    `yarn start`
>    `a` – both Android simulator and a real Android device, Storybook works well on both options but does not in a browser.
> 4. On `yarn storybook` the answer:
> 
> ```
> yarn run v1.22.4
> $ start-storybook -p 7007
> info => Loading presets
> info => Loading custom manager config.
> ╭────────────────────────────────────────────────────╮
> │                                                    │
> │   Storybook 5.3.18 started                         │
> │   4.39 s                                           │
> │                                                    │
> │    Local:            http://localhost:7007/        │
> │    On your network:  http://192.168.0.104:7007/    │
> │                                                    │
> ╰────────────────────────────────────────────────────╯
> ```
> 
> **From the tutorial it should be this**
> <img alt="from tutorial" width="356" src="https://user-images.githubusercontent.com/63786160/82722658-62ecb780-9cd1-11ea-97ad-5e7a3d04ff3c.png">
> 
> **But I have this result on any projects and environments I tried**
> <img alt="error" width="843" src="https://user-images.githubusercontent.com/63786160/82723362-12785880-9cd7-11ea-87e9-a2142f46cbe9.png">
> 
> **Environment Info:**
> System:
> OS: macOS Mojave 10.14.6
> CPU: (8) x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
> Binaries:
> Node: 12.16.3 - /usr/local/bin/node
> Yarn: 1.22.4 - /usr/local/bin/yarn
> Browsers:
> Chrome: 81.0.4044.138
> Firefox: 76.0
> Safari: 12.1.2
> npmPackages:
> @storybook/addon-actions: ^5.3.18 => 5.3.18
> @storybook/addon-links: ^5.3.18 => 5.3.18
> @storybook/addon-ondevice-actions: ^5.3.18 => 5.3.18
> @storybook/addons: ^5.3.18 => 5.3.18
> @storybook/react-native: ^5.3.18 => 5.3.18
> @storybook/react-native-server: ^5.3.18 => 5.3.18
> 
> Also, I tried with the older 5.2.1 Storybook version, an Android simulator, and an Android phone via USB, Ubuntu, manual server setup, but the same error.
> 
> I also set up React Storybook Tutorial for a try to see if there is this error there. No problem, all works properly with React.

I found the solution. Same problem happened with me.

Please start your story book with this argument -> `-h 0.0.0.0`. The whole command would be -> `start-storybook -h 0.0.0.0`,
You can also change it in your "package.json" file, so that you can just use "yarn storybook".

After the server starts, if the browser says "This site can't be reached", for the ip `0.0.0.0`, just open the server with your `local ip address`, for example, my address is `192.168.45.150`, so I would open it with `192.168.45.150:7007`, (I have my server running on `7007` port, if you have it bind to some other port, please use that).

There seems to be some network issue with storybook and it runs only on your localhost if the command `start-storybook` is used without "host" argument. As a result the emulator or your device can't find the storybook server, hence no stories are displayed. 
You can see below the screenshot, I can see the stories now.

![image](https://user-images.githubusercontent.com/46626546/93014697-cb3d0d00-f5d0-11ea-9ed0-c5e11fd2b837.png)

Also no need of writing any port or host in `getStorybookUI({})`. It will work without it.

Found the major hint in [this](https://github.com/storybookjs/storybook/issues/6674#issuecomment-511818022) reply. 


**EDIT 1:**
Okay.. sometimes we do need to mention port and host in  `getStorybookUI({})`.

**Comment by dannyhw** [comment]
@Shhhhhreyas thanks for sharing, hopefully this will help others with the same problem 🙌

**Comment by JeffGuKang** [comment]
I am using a custom script for node env as `STORYBOOK=true react-native start`
And `--reset-cache` option is necessary  when I run `STORYBOOK=true react-native start` if I ran `react-native start` before.

```
const EntryApp: NavigationContainer =
  process.env.STORYBOOK
    ? require('./storybook').default
    : createAppContainer(RootNavigator)
```

**Comment by EsackN** [comment]
> > Hello! I had a problem with loading Storybook web explorer, it doesn't load the left menu. But on a device Storybook works well. I met the problem both during Intro React Native Storybook Tutorial and in my bare workflow project. It just doesn't build a web version.
> > **Steps**
> > 
> > 1. In a terminal to a project folder:
> >    `npx -p @storybook/cli sb init --type react_native`
> >    `yarn add -D @storybook/addon-ondevice-actions`
> > 2. In storybook/rn-addons.js
> >    `import '@storybook/addon-ondevice-actions/register';`
> > 3. In the terminal I tried this and reverse order:
> >    `yarn storybook`
> >    `yarn start`
> >    `a` – both Android simulator and a real Android device, Storybook works well on both options but does not in a browser.
> > 4. On `yarn storybook` the answer:
> > 
> > ```
> > yarn run v1.22.4
> > $ start-storybook -p 7007
> > info => Loading presets
> > info => Loading custom manager config.
> > ╭────────────────────────────────────────────────────╮
> > │                                                    │
> > │   Storybook 5.3.18 started                         │
> > │   4.39 s                                           │
> > │                                                    │
> > │    Local:            http://localhost:7007/        │
> > │    On your network:  http://192.168.0.104:7007/    │
> > │                                                    │
> > ╰────────────────────────────────────────────────────╯
> > ```
> > 
> > 
> > **From the tutorial it should be this**
> > <img alt="from tutorial" width="356" src="https://user-images.githubusercontent.com/63786160/82722658-62ecb780-9cd1-11ea-97ad-5e7a3d04ff3c.png">
> > **But I have this result on any projects and environments I tried**
> > <img alt="error" width="843" src="https://user-images.githubusercontent.com/63786160/82723362-12785880-9cd7-11ea-87e9-a2142f46cbe9.png">
> > **Environment Info:**
> > System:
> > OS: macOS Mojave 10.14.6
> > CPU: (8) x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
> > Binaries:
> > Node: 12.16.3 - /usr/local/bin/node
> > Yarn: 1.22.4 - /usr/local/bin/yarn
> > Browsers:
> > Chrome: 81.0.4044.138
> > Firefox: 76.0
> > Safari: 12.1.2
> > npmPackages:
> > @storybook/addon-actions: ^5.3.18 => 5.3.18
> > @storybook/addon-links: ^5.3.18 => 5.3.18
> > @storybook/addon-ondevice-actions: ^5.3.18 => 5.3.18
> > @storybook/addons: ^5.3.18 => 5.3.18
> > @storybook/react-native: ^5.3.18 => 5.3.18
> > @storybook/react-native-server: ^5.3.18 => 5.3.18
> > Also, I tried with the older 5.2.1 Storybook version, an Android simulator, and an Android phone via USB, Ubuntu, manual server setup, but the same error.
> > I also set up React Storybook Tutorial for a try to see if there is this error there. No problem, all works properly with React.
> 
> I found the solution. Same problem happened with me.
> 
> Please start your story book with this argument -> `-h 0.0.0.0`. The whole command would be -> `start-storybook -h 0.0.0.0`,
> You can also change it in your "package.json" file, so that you can just use "yarn storybook".
> 
> After the server starts, if the browser says "This site can't be reached", for the ip `0.0.0.0`, just open the server with your `local ip address`, for example, my address is `192.168.45.150`, so I would open it with `192.168.45.150:7007`, (I have my server running on `7007` port, if you have it bind to some other port, please use that).
> 
> There seems to be some network issue with storybook and it runs only on your localhost if the command `start-storybook` is used without "host" argument. As a result the emulator or your device can't react the storybook server, hence no stories are displayed.
> You can see below the screenshot, I can see the stories now.
> 
> ![image](https://user-images.githubusercontent.com/46626546/93014697-cb3d0d00-f5d0-11ea-9ed0-c5e11fd2b837.png)
> 
> Also no need of writing any port or host in `getStorybookUI({})`. It will work without it.
> 
> Found the major hint in [this](https://github.com/storybookjs/storybook/issues/6674#issuecomment-511818022) reply.
> 
> **EDIT 1:**
> Okay.. sometimes we do need to mention port and host in `getStorybookUI({})`.

**Comment by EsackN** [comment]
@Shhhhhreyas I  followed the steps as you mentioned, still getting the same error.

**Comment by Shhhhhreyas** [comment]
> @Shhhhhreyas I followed the steps as you mentioned, still getting the same error.

@EsackN
Did you add `host` in `getStorybookUI({})`. Example -> `getStorybookUI({host:"192.168.43.66"});` (your local IP or the one which storybook mentions in the box after starting)

This function is used in `index.js` inside `storybook` folder in your project.

Also you have to start the app in your phone, without it stories won't show.

**Comment by EsackN** [comment]
@Shhhhhreyas Yes, I added host in getStorybookUI({}). Also please clarify with the following. 

I refered this link: https://stackoverflow.com/questions/60696774/how-can-i-build-my-react-native-storybook-to-web

IS that right?

**Comment by Shhhhhreyas** [comment]
@EsackN 
I think the link maybe correct I didn't setup it this way. It seems long and lot of stuff has to be done.

I used the below steps ->

`# Add Storybook:
npx -p @storybook/cli sb init --type react_native`

(During Storybook's install process, you'll be prompted to install react-native-server, do so as this package will help out immensely throughout the tutorial.)

We'll also want to add another package and make a change to storybook/rn-addons.js to allow the actions (you'll see them in action later in the tutorial) to be logged correctly in the Storybook UI.

Run the following command:

`yarn add -D @storybook/addon-ondevice-actions`

Change storybook/rn-addons.js to the following:

`// storybook/rn-addons.js`
`import '@storybook/addon-ondevice-actions/register';`


This was all I did. Source - https://www.learnstorybook.com/intro-to-storybook/react-native/en/get-started/

And then that `-h 0.0.0.0` thing.

Try this and let me know.

**EDIT:** The link given by you is not for storybook server. It is for react native web. Don't follow that.

**Comment by dannyhw** [comment]
FYI: the cli is currently a bit broken for react native. I've already updated this but its in alpha. If you need to setup a new project I recommend you use the @next version of the cli 

`npx -p @storybook/cli@next sb init --type react_native`

Also please bear in mind that the screenshots in the storybook tutorial actually show the web version of the tutorial and you won't see the same thing on your side (it will be kind of similar). I'm currently working with the maintainer of the tutorial to improve it however it will take some time to update everything.

**Comment by EsackN** [comment]
No luck guys @dannyhw and @Shhhhhreyas.  It would be grateful, if anyone can connect with me on hangouts to resolve this issue.
esack18nsne@gmail.com

**Comment by Shhhhhreyas** [comment]
@EsackN Sent you a invite from 'shreyascool.sh@gmail.com'.

**Comment by dannyhw** [comment]
@EsackN if you still need help I'm usually active on the react native channel in the storybook Discord. I can't do a voice call until the evening (UTC+1) since I have work.

**Comment by EsackN** [comment]
@Shhhhhreyas Thank you so much. @dannyhw Thanks for the words. @Shhhhhreyas Helped me and the issue got fixed,

**Comment by EsackN** [comment]
@Shhhhhreyas @dannyhw Now facing another issue. When i click any stories on emulator, web is not previewing the same. 

 "@storybook/addon-actions": "^5.3.19",
    "@storybook/addon-knobs": "^6.0.22",
    "@storybook/addon-links": "^5.3.19",
    "@storybook/addon-ondevice-actions": "^5.3.21",
    "@storybook/addon-ondevice-knobs": "^5.3.19",
    "@storybook/addon-ondevice-notes": "^5.3.19",
    "@storybook/addons": "^5.3.19",
    "@storybook/react-native": "^5.3.19",
    "@storybook/react-native-server": "^5.3.19",

![storybook-working-web](https://user-images.githubusercontent.com/41628212/94427608-ffa5f100-01ac-11eb-8c9d-518b9f922ed4.jpg)

**Comment by dannyhw** [comment]
@EsackN that's not how it works, the server doesn't show any preview. You'll only see the selected story change. To see it on the web you need to run the on-device UI on the web. Like `yarn web` if on expo.

**Comment by Jonovono** [comment]
@EsackN how did you resolve your first issue?

**Comment by tommasini** [comment]
Hey guys, I'm here to relate I can't solve the first issue, my stories list are infinitely loading, I did every single step wich one of you said, and my final result is this one:
![image](https://user-images.githubusercontent.com/46944231/95441411-058c9680-0952-11eb-87c6-042646302f36.png)

-If anyone could help me I'm available to do a voice call! I don't know what I can do, I did everything:
- I have the emulator working on pc and also a physical device.
- I created a new reac native app for try this, because I was trying configure this on a private repository.
- I followed the getting started react native storybook tutorial, didn't work.
- I tried to write the host on my package.json -h [localhost] and tried also [0.0.0.0] then open my local network.
- Also I follow the instruction of, first running my storybook react native server and then run the app and open it,
- I changed also my getStorybookUI({}] with my local network host and port 7007, also tried with asyncStorage false and null (crying at this time ahah)
- I created a new storie following the next step of the tutorial and changed the configure of storybook/index.js for that path, 
because the problem could be of the stories automatically created of the tutorial, also didn't work.
- I changed the version of storybook it was 5.3.23 and I tried 5.3.21, like @dannyhw said in some abose answer.


Can some one help me? I don't know more what I can do.

Best regards,
Keep healthy,

Tomás Santos

**Comment by Shhhhhreyas** [comment]
@tommasini
Let's connect.. Ping me on hangouts - shreyascool.sh@gmail.com.

**Comment by dannyhw** [comment]
@tommasini when I mention storybook version you can use the latest for `@storybook/react-native`,  `@storybook/react-native-server`  and `@storybook/ondevice-[addon-name]` but when it comes to storybook packages from the main repo just make sure they aren't using version 6.0 and above. When unsure you can just use the version as `^5.3` which will get you the latest 5.3.x version.

Likely you just need to find the right combination of configs here. Make sure that if you change the host on your storybookUI() config that you set the same host and port on the storybook server command in package json.

**Comment by tommasini** [comment]
Hello!!
Thanks to @Shhhhhreyas I was able to configurate the Storybook. We did a call and we can notice I wasn't importing in my app.js my storybook/index.js and rendering that component I will show what it was missing on a image below:
![image](https://user-images.githubusercontent.com/46944231/95576918-2d026280-0a29-11eb-8301-23a15da6e086.png)

- Just need to import the storybook/index.js to your app.
- Comment all code you are rendering
- Render the component you just imported

Hope you all get the problem done.

Thanks @Shhhhhreyas 

Best regards,

Tomás Santos

**Comment by dannyhw** [comment]
@Shhhhhreyas thanks for going out of your way to help people in the community, its great to see 🙂.

**Comment by Shhhhhreyas** [comment]
> Hello!!
> Thanks to @Shhhhhreyas I was able to configurate the Storybook. We did a call and we can notice I wasn't importing in my app.js my storybook/index.js and rendering that component I will show what it was missing on a image below:
> ![image](https://user-images.githubusercontent.com/46944231/95576918-2d026280-0a29-11eb-8301-23a15da6e086.png)
> 
> * Just need to import the storybook/index.js to your app.
> * Comment all code you are rendering
> * Render the component you just imported
> 
> Hope you all get the problem done.
> 
> Thanks @Shhhhhreyas
> 
> Best regards,
> 
> Tomás Santos

Your most welcome @tommasini 😄. Just happy that your problem got solved.

Happy `coding` 😊

**Comment by Shhhhhreyas** [comment]
> @Shhhhhreyas thanks for going out of your way to help people in the community, its great to see 🙂.

Thank you @dannyhw 😅.
Just trying to help people. 😃

**Comment by MarceloPrado** [comment]
Hello everyone 👋🏻 

I tried everything mentioned but the **only solution was to use my network IP address instead of "0.0.0.0"**. To avoid hardcoding the address in two places (`package.json` and `storybook.js`), I created a simple bash script automates this task for us.

**The final solution looks like this:**

**1**. In `package.json`, under `scripts`, add the following line: `"storybook": "tasks/storybook.sh",`

**2.** Add the bash script below under a `tasks` folder. Name it `storybook.sh`.
```bash
#!/bin/bash
set -euo pipefail

# Returns computer IP address
# This is used in config/storybook, fixing the infinite loading sidebar bug
# See: https://github.com/storybookjs/react-native/issues/55
host=$(ifconfig en0 | grep inet | grep -v inet6 | awk '{print $2}')

json="{\"runStorybook\": true, \"host\": \"$host\", \"port\": 7007 }"
echo $json > config/storybook/config.json

printf "\n🚨 ️Don't forget to run \"yarn android\" in a new shell once the server is up!\n\n"
start-storybook -p 7007 -h $host
```
**2.1** Don't forget to give the right permission to the script: `chmod +x tasks/storybook.sh`

**3.** Modify your `storybooks.js` to support the new format:
```js
import { getStorybookUI, configure } from '@storybook/react-native';
import { AppRegistry } from 'react-native';

import { name as appName } from '../../app.json';

import { host, port } from './config.json';
import { loadStories } from './storyLoader';

configure(() => {
  loadStories();
}, module);

const StorybookUIRoot = getStorybookUI({
  asyncStorage: null,
  // `host` and `port` are dynamically generated from storybook.sh
  host,
  port,
});

export default StorybookUIRoot;

AppRegistry.registerComponent(appName, () => StorybookUIRoot);

```

**4.** **DONE**! Execute `yarn storybook` and, in a new shell, `yarn android` or `yarn ios`.
![Screenshot 2020-10-25 at 16 31 57](https://user-images.githubusercontent.com/8047841/97116961-9cda4380-16df-11eb-94d9-6854e4c81035.png)

**Comment by krystofmatejka** [comment]
I was struggling for 2 days to make this work.

In my case, the problem was in networking.

Run storybook on IP 0.0.0.0 and test with some tool (for example nmap -p 7007 YOUR_NETWORK_IP) if is port open on the network IP. If is not try to run storybook with **--ci**.

**Comment by dannyhw** [comment]
@kxmatejka sorry to hear that, if you have issues like this is the future feel free to ask questions on the [Discord](https://discord.gg/sMFvFsG) I'm active there and I can help with these types of issues :) 

Another thing is that the server isn't needed for a normal use cases and I would advise people to use the on device ui without the server for most cases.

**Comment by krystofmatejka** [comment]
@dannyhw I really appreciate your work. Now everything runs smoothly. But the setup process was a nightmare. Every tutorial/guide/blog shows the same: run init and there you go. Then if something failed, it is almost impossible for the newcomer to find out what is wrong and how to fix it.

It would be nice add to docs some scheme with an explanation where should run what and where to look if is something broken.

For example, I still fully don't get the relationship between storybook server - react native server - application

**Comment by dannyhw** [comment]
@kxmatejka I totally understand where you're coming from and a big part of the issue is that the official tutorial is inaccurate. I've done some work to help get it changed but it's been hard to get the right people together to make the changes needed. 

I'm hoping to solve a lot of issues in 6.0 that will make the setup much easier.

Most likely server won't be part of 6.0 since it just causes confusion and it also makes the upgrade to 6.0 more complicated.

Lastly if it would be beneficial I could make a quick blog post or something to help explain the setup and try to clear up some things.

**Comment by krystofmatejka** [comment]
@dannyhw that will be really nice. I really want to understand how it works, but there aren't any resources.

**Comment by MarceloPrado** [comment]
I ended up writing a bash script that automates all of this, fixing my issues. However, it wasn't trivial at all. Let me know if you need any help @dannyhw, I believe the 6.0 will be a great addition to react-native 🙂

**Comment by Man-RP** [comment]
struggling the same issue, tried the 0.0.0.0 ip thing - did not work.
the shel script is not working for me too :S
![image](https://user-images.githubusercontent.com/46972103/103461646-a360e900-4d28-11eb-84da-d9975a474b02.png)

getting:
![image](https://user-images.githubusercontent.com/46972103/103461694-ede26580-4d28-11eb-86aa-c617a258afff.png)

edit: @dannyhw helped me to solve it. thanks a lot!

iv'e got another problem with my device's app screen:
![WhatsApp Image 2021-01-02 at 19 14 48](https://user-images.githubusercontent.com/46972103/103462719-e5415d80-4d2f-11eb-8e45-e213d62d6dc5.jpeg)

**Comment by jo-finndian** [comment]
> You should run storybook server and then after run the react native app then if you refresh the webpage for react native server the stories should load.

Omg, I've been banging my head against the wall for hours. I feel like such a dolt. Thank you!

**Comment by Yasin-Ali** [comment]
I did everything above and it is still not working :-/

**Comment by dannyhw** [comment]
@Yasin-Ali If you share more details I could help you get it working. What's the server command look like in your package.json and the getstorybookui arguments you are using.

**Comment by ivancho2** [comment]
in the latest versions of Storybook the user is asked to configure the input import manually.

see [https://www.npmjs.com/package/@storybook/react-native](https://www.npmjs.com/package/@storybook/react-native)

in my case for template project without expo I had to add the following line to /index.js

`
export {default} from "./storybook";
`

**Comment by anhdn** [comment]
<img width="912" alt="note" src="https://user-images.githubusercontent.com/864385/124574202-71197b00-de74-11eb-95e3-bd55d6909b79.png">
I read it on this repo document. Maybe we can not run storybook on web independently. We have to run web and simulator at the same time.

**Comment by dannyhw** [comment]
@anhdn 
> I read it on this repo document. Maybe we can not run storybook on web independently. We have to run web and simulator at the same time.

Yes the storybook server is not a web storybook for react native it's just a server for controlling the device.

**Comment by RitikaGupta8734** [comment]
I am also facing the same issue. Above solution didn't worked for me. Can someone help? 
<img width="956" alt="Screenshot 2021-07-09 at 7 19 50 PM" src="https://user-images.githubusercontent.com/43800283/125088311-eb533500-e0ea-11eb-9fcb-11fc9371edbf.png">
<img width="371" alt="Screenshot 2021-07-09 at 7 20 23 PM" src="https://user-images.githubusercontent.com/43800283/125088367-f7d78d80-e0ea-11eb-9daf-570fbd2b82f3.png">

**Comment by RitikaGupta8734** [comment]
Following steps will work in case you are able to load stories on your emulator and control them via on-device navigator, preview and addons and just `want to add the server control along with on device one`.

Advantages: 
1. Storybook server is used to control the component visible on the device via a web ui.
2. It is useful to control multiple devices at once and compare them at the same time.
3. It provides far better UI then the on-device one.

Steps:
1. If you have installed storybook then your `package.json` file will have `"storybook": "start-storybook -c .storybook` in scripts
2. Now you will run `npm run storybook` and the server will open in which the stories will `load infinitely`.
3. Now before starting your app make the following changes.
4. Copy the host and the port part from the terminal where you ran `npm run storybook`. It will look like this
╭─────────────────────────────
│                                                     
│   Storybook 5.3.23 started                          
│   5.13 s                                            
│                                                     
│    Local:            http://localhost:7007/    
│    On your network:  http://192.168.137.15:7007/    
│                                                     
╰─────────────────────────────
host-> `192.168.137.15`(copy your host) and port -> `7007`
5. Now specify these at exactly two places : 
    1. `package.json` -> `"storybook": "start-storybook -c .storybook -p 7007 -h 192.168.137.15" `  and
    2.  `storybook/stories/index.js` ->  `const StorybookUIRoot = getStorybookUI({ port: 7007, host: '192.168.137.15' })`
6. Now stop that `npm run storybook` command by pressing `ctrl +c` and run it again
7. If you did it all right then it will look like below , else recheck if you followed all the steps.
╭─────────────────────────────
│                                                     
│   Storybook 5.3.23 started                          
│   5.13 s                                            
│                                                     
│    Local:            http://192.168.137.15:7007/    
│    On your network:  http://192.168.137.15:7007/    
│                                                     
╰─────────────────────────────
8. Now start you app as usual: `npm start' and all in other terminals (keep the `npm run storybook` running) and reload the storybook server. You will see it working now.

**Comment by RitikaGupta8734** [comment]
Following steps will work in case you are able to load stories on your emulator and control them via on-device navigator, preview and addons and just `want to add the server control along with on device one`.
I faced this problem and @dannyhw helped me.

Advantages: 
1. Storybook server is used to control the component visible on the device via a web ui.
2. It is useful to control multiple devices at once and compare them at the same time.
3. It provides far better UI then the on-device one.

Steps:
1. If you have installed storybook then your `package.json` file will have `"storybook": "start-storybook -c .storybook` in scripts
2. Now you will run `npm run storybook` and the server will open in which the stories will `load infinitely`.
3. Now before starting your app make the following changes.
4. Copy the host and the port part from the terminal where you ran `npm run storybook`. It will look like this

```
╭─────────────────────────────
│                                                     
│   Storybook 5.3.23 started                          
│   4.18 s                                            
│                                                     
│    Local:            http://localhost:7007/     
│    On your network:  http://192.168.137.15:7007/    
│                                                     
╰────────────────────────────
```

host-> `192.168.137.15`(copy your host) and port -> `7007`
5. Now specify these at exactly two places : 
```
1. `package.json` -> `"storybook": "start-storybook -c .storybook -p 7007 -h 192.168.137.15" `  and
2. `storybook/stories/index.js` ->  `const StorybookUIRoot = getStorybookUI({ port: 7007, host: '192.168.137.15' })`
```
6. Now stop that `npm run storybook` command by pressing `ctrl +c` and run it again
7. If you did it all right then it will look like below , else recheck if you followed all the steps.
```
╭─────────────────────────────
│                                                     
│   Storybook 5.3.23 started                          
│   5.13 s                                            
│                                                     
│    Local:            http://192.168.137.15:7007/    
│    On your network:  http://192.168.137.15:7007/    
│                                                     
╰─────────────────────────────
```
8. Now start you app as usual: `npm start` and all in other terminals (keep the `npm run storybook` running) and reload the storybook server. You will see it working now.

**Comment by abhishek-netizen** [comment]
please refer above post by @RitikaGupta8734 .  its working for me

**Comment by kenchoong** [comment]
> I found there is an issue when using a real android device with expo. It can be resolved by specifying your port and host explicitly
> 
> When you start up the server you should see something with your IP here
> On your network: http://xxx.xxx.x.xx:7007/
> 
> Take this IP and set it explicitly as the host and then the storybookui should be able to reach the server
> 
> In code
> 
> ```
> const StorybookUIRoot = getStorybookUI({
>   host: "192.111.1.11", // replace this ip address with your local ip address
>   port: "7007"
> });
> ```
> 
> In package.json `"storybook": "start-storybook -p 7007 -h 192.111.1.11"` (replace this ip address with your local ip address
> port: "7007")
> 
> This should resolve the connection issues
> 
> This might also help some people if the above isn't working
> 
> `adb reverse tcp:7007 tcp:7007`

Hey wanna ask, do u run your expo on device by scan qr code or connect with cable? Cause now both android, ios I scan with Expo Qr code, both cant connect with storybook. 

What I have done: 

1. set the storybook UI like this
```ts
const StorybookUIRoot = getStorybookUI({
  host:'173.68.789.25',  // this is get bro the Storybook output 
  port:'7007',
  asyncStorage: null
});
```

2. expo start 
3. iOS scan expo QR code 
4. android scan expo QR code 
5. yarn storybook 

Outcome now: 
1. Android and iOS both having the Storybook present 
2. In storybook, the menu is loading forever. 

Not sure what happened here 😢

It look likes  the Android iOS device cant connect with the Storybook server. 

Let me know if have any suggestion

**Comment by dannyhw** [comment]
@kenchoong

Make sure you aren't connecting over the "tunnel" connection mode in expo, try to use local LAN network option. 

Also make sure you changed the server port in your package.json to the same as your getStorybookUI one.

**Comment by kenchoong** [comment]
> @kenchoong
> 
> Make sure you aren't connecting over the "tunnel" connection mode in expo, try to use local LAN network option.

Yes, I use LAN connection, I tried both "Scan QRCode" and also connect with cable.. both cant connect(forever loading in `http://172.30.48.1:7007/?path=/story/*`). 

In package.json, this is my script. 

```
"storybook": "start-storybook -p 7007 -h 172.30.48.1",
```

Then getStorybookUI is this. 

```
const StorybookUIRoot = getStorybookUI({
  host:'172.30.48.1',
  port:'7007',
  asyncStorage: null
});
```

and `172.30.48.1` is get from `yarn storybook` output for the 1st time

I also `yarn storybook` first, then only `expo start` 

In android device, storybook UI is presented, just cant connect in `localhost:7007` 

But still no luck 😢

**Comment by dannyhw** [comment]
@kenchoong have you tried `adb reverse tcp:7007 tcp:7007` ?

**Comment by kenchoong** [comment]
> @kenchoong have you tried `adb reverse tcp:7007 tcp:7007` ?

errm. about this, how can I run this command ya? Sorry, I just getting started

**Comment by kenchoong** [comment]
> @kenchoong have you tried `adb reverse tcp:7007 tcp:7007` ?

Yes, 

1. yarn storybook 
2. expo start
3. adb reverse tcp:7007 tcp:7007

run each command after another. What I got now: 

1. android got the preview of Storybook 
2. http://172.30.48.1:7007/?path=/story/* in this path, storybook menu infinite loading. 

Connected using cable and Scan QRcode as well , weird..

**Comment by dannyhw** [comment]
try changing it to this
```
const StorybookUIRoot = getStorybookUI({
  host:'172.30.48.1',
  port:7007,
  asyncStorage: null
});
```
I don't think that the port should be a string.

**Comment by kenchoong** [comment]
> try changing it to this
> 
> ```
> const StorybookUIRoot = getStorybookUI({
>   host:'172.30.48.1',
>   port:7007,
>   asyncStorage: null
> });
> ```
> 
> I don't think that the port should be a string.

Yes, this is one of the reason. 

And state here for future references as well, the reason of this is because `host:'172.30.48.1'` this value of host(I got from the output after yarn storybook) is wrong. After I restart my computer, yarn storybook again, it gives a new value which is `192.156.1.2` (a totally new value), replace that, the whole thing works which means sidebar menu can see the component, it works with both  "scan expo qr code" and "connect device with cable", both in android and ios device works. Struggle on this for 1 WHOLE DAY. 🤣😂

**Comment by alutz33** [comment]
@dannyhw @kennethtruong WORKED FOR ME! Just needed to add the host and port. What a waste of a day!

**Comment by jordanlang410** [comment]
I am still struggling with the sidebar not loading and I have completed everything exactly as explained by @kenchoong.  I'm not sure what I would be doing wrong at this point?  Does it matter what server expo is running on?

**Comment by dannyhw** [comment]
> I am still struggling with the sidebar not loading and I have completed everything exactly as explained by I'm not sure what I would be doing wrong at this point? Does it matter what server expo is running on?

@jordanlang410 The important thing is having the same ip and port for both the storybook script in package.json and also in getStorybookUI. Also make sure its the right ip address for your machine. It should be the local network address, this ip will be shown when you initially run the server. This solution has been shown to work multiple times so I imagine you just need to double check your setup and double check the IP address. 

In package.json

"storybook": "start-storybook -p 7007 -h 172.1.1.1",

and getStorybookUI

const StorybookUIRoot = getStorybookUI({
  host:'172.1.1.1',
  port:'7007',
  asyncStorage: null
});

**Comment by kenchoong** [comment]
> I am still struggling with the sidebar not loading and I have completed everything exactly as explained by @kenchoong. I'm not sure what I would be doing wrong at this point? Does it matter what server expo is running on?

add this to your package.json 

```
"storybook": "start-storybook -p 7007 -h 192.156.1.2",
```
The value of the host should start with 192, other than that, certainly it is wrong. Restart your computer, and get the new IP.

**Comment by jordanlang410** [comment]
I figured it out.  I did not follow instructions properly and failed to put the export {default} from './storybook'; into my App.js file.  You all rock @kenchoong and @dannyhw.  Your earlier posts saved me a ton of time.

**Comment by Dane-Mutch** [comment]
Leaving this here in case it helps anybody. We use the new JSX transform as described [here](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) and were facing an infinitely loading sidebar. Setting the `experimentalImportSupport` option in `metro.config.js` to true fixed our issue.

**Comment by dannyhw** [comment]
I'm just locking this to avoid unnecessary noise, I believe that there are already solutions and discussing further isn't necessary. I'll leave the issue open for discoverability and to help others with the same problem.

**Comment by dannyhw** [comment]
as this issue is very old I believe we can close this since the server was completely replaced in v7
